### ISSUE TITLE
Support Updating Individual Array Elements

### ISSUE THREAD

**Issue by philip-alldredge** [diesel-rs/diesel#1681]
## Problem Description
As far as I am aware, Diesel does not provide a mechanism for referencing individual array elements inside of the SQL queries.

### What are you trying to accomplish?
Updating a single element in an array column when working with a Postgres database. I'd like to generate a query like:

```
UPDATE table_with_array
SET my_array[2] = 5.0;
```
Using syntax like:
`update(table_with_array.set(my_array.index(1).eq(5.0)).execute(db_connection);`

**Comment by sgrif** [comment]
This is a reasonable feature request, but isn't likely to get added in the near future. I'd recommend using [`sql_query`](http://docs.diesel.rs/diesel/fn.sql_query.html) for this.

**Comment by philip-alldredge** [comment]
Okay. Thanks! It's only a minor inconvenience. Everything else works great.

**Comment by sgrif** [comment]
Glad to hear it. :smile:

**Comment by sgrif** [comment]
I actually don't think this would be super hard to implement if someone were interested in tackling it (being able to select a single element of an array would be a good first step. I think this might make sense as an issue for @galuhsahid or @yamishi13 in a few weeks.

**Comment by weiznich** [comment]
Just to write it somewhere:
I think this should be done in a similar manner like the implementation of  numeric operations in diesel for various types. This means in detail that instead of providing a index method we should use the `Index` trait from the std-lib and then return a custom wrapper for codegen from there. (Just like for Add for example)

**Comment by Justice4Joffrey** [comment]
I've just had a look at how this might be implemented. 

The docs suggest arrays containing `NULL` aren't supported. Afaict that's a bad fit for Postgres arrays which can always contain `NULL` elements.

e.g. demonstrated by this somewhat surprising behaviour -

```
db=# CREATE TABLE posts (id SERIAL PRIMARY KEY, tags TEXT[] NOT NULL);
db=# insert into posts (tags) values (array['cool', 'great']);
db=# update posts set tags[5] = 'incredible';
db=# select * from posts;
+----+-----------------------------------+
| id |               tags                |
+----+-----------------------------------+
|  1 | {cool,great,NULL,NULL,incredible} |
+----+-----------------------------------+
```

Any Array which ends up containing `NULL`s will break deserialization with `DeserializationError(UnexpectedNullError)`. Therefore, offering the `index()` method is only going to make this footgun a lot easier to trigger.

IMO this means Postgres `Array<ST>` should always map to a `Vec<Option<T>>`. Clearly quite a disruptive change though.

**Comment by weiznich** [comment]
@Justice4Joffrey Good point. Instead of changing the mapping so that `Array<ST>` is only compatible with `Vec<Option<T>>` I would rather like to change the generated schema, so that it reads `Array<Nullable<ST>>`, which automatically enforces the correct mapping and makes that more obvious.  Additionally this would enable users to opt into arrays with non nullable elements manually for cases they are really sure that something will not contain any null value.

Mind submitting a PR for this? It should be as easy as

* Modifying the printing code here: https://github.com/diesel-rs/diesel/blob/95d1eabf10a5f4318bf7e2490bc626e02facefd1/diesel_cli/src/infer_schema_internals/data_structures.rs#L31-L54
* Add a new test case here: https://github.com/diesel-rs/diesel/blob/95d1eabf10a5f4318bf7e2490bc626e02facefd1/diesel_cli/tests/print_schema.rs
* Add a changelog entry here: https://github.com/diesel-rs/diesel/blob/master/CHANGELOG.md#changed

**Comment by Justice4Joffrey** [comment]
Sure - #3056

**Comment by weiznich** [comment]
Closed by #3093
